File: clients/dart/test/sync_page_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:cliptown_interfaces/cliptown_interfaces.dart';
import 'package:test/test.dart';

Map<String, Object?> _map(Object? value) =>
    (value! as Map<Object?, Object?>).cast<String, Object?>();

String _timestamp(DateTime value) =>
    value.toUtc().toIso8601String().replaceFirst('.000Z', 'Z');

void main() {
  test('final page advances cursor and preserves tombstone', () {
    final source = _map(
      jsonDecode(File('../../fixtures/sync-page.json').readAsStringSync()),
    );
    final items = source['items']! as List<Object?>;
    final tombstone = ClipEnvelope.fromJson(_map(items.single));
    final cursor = _map(source['next_cursor']);

    tombstone.validate();
    expect(source['has_more'], isFalse);
    expect(cursor['cursor'], 'server-sequence:42');
    expect(cursor['server_sequence'], 42);
    expect(tombstone.deleted, isTrue);
    expect(tombstone.pinned, isFalse);
    expect(tombstone.blindTerms, isEmpty);
    expect(tombstone.optInEmbedding, isNull);

    final roundTrippedTombstone = tombstone.toJson()
      ..['created_at'] = _timestamp(tombstone.createdAt)
      ..['updated_at'] = _timestamp(tombstone.updatedAt);
    final roundTripped = <String, Object?>{
      'items': <Object?>[roundTrippedTombstone],
      'next_cursor': <String, Object?>{
        'cursor': cursor['cursor'],
        'server_sequence': cursor['server_sequence'],
      },
      'has_more': source['has_more'],
    };

    expect(roundTripped, equals(source));
  });
}

File: clients/typescript/test/sync-page.test.ts
import assert from 'node:assert/strict';
import {readFileSync} from 'node:fs';
import test from 'node:test';

import {
  validateClipEnvelope,
  type ClipEnvelope,
  type SyncCursor,
} from '@cliptown/interfaces';

interface SyncPageFixture {
  items: ClipEnvelope[];
  next_cursor: SyncCursor;
  has_more: boolean;
}

test('final page advances cursor and preserves tombstone', () => {
  const source = JSON.parse(
    readFileSync(new URL('../../../fixtures/sync-page.json', import.meta.url), 'utf8'),
  ) as SyncPageFixture;

  assert.equal(source.has_more, false);
  assert.equal(source.next_cursor.cursor, 'server-sequence:42');
  assert.equal(source.next_cursor.server_sequence, 42);
  assert.equal(source.items.length, 1);

  const tombstone = source.items[0];
  assert.ok(tombstone);
  validateClipEnvelope(tombstone);
  assert.equal(tombstone.deleted, true);
  assert.equal(tombstone.pinned, false);
  assert.deepEqual(tombstone.blind_terms, []);
  assert.equal(tombstone.opt_in_embedding, null);

  assert.deepEqual(JSON.parse(JSON.stringify(source)), source);
});
